src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_SIZE = 10;
+    private static final double GROW_FACTOR = 1.5;
+
+    private T[] elements;
+    private int size;
+
     @Override
     public void add(T value) {
-
+        addCellForNewElement(value);
+        elements[size - 1] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        int prevSize = size;
+        addCellForNewElement(value);
+        validateIndex(index);
+        if (index < prevSize) {
+            System.arraycopy(elements, index, elements, index + 1, prevSize - index);
+        }
+        elements[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            throw new NullPointerException("Input list is null");
+        }
+        if (list.size() == 0) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        final T value = elements[index];
+        size--;
+        if (index < size) {
+            System.arraycopy(elements, index + 1, elements, index, size - index);
+        }
+        elements[size] = null;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] == element)
+                    || (elements[i] != null && elements[i].equals(element))) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("No such element in the list: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addCellForNewElement(T value) {
+        if (elements == null) {
+            elements = createArray(value, INITIAL_SIZE);
+        }
+        size++;
+        if (size > elements.length) {
+            T[] newArray = createArray(value, (int) (elements.length * GROW_FACTOR));
+            System.arraycopy(elements, 0, newArray, 0, elements.length);
+            elements = newArray;
+        }
+    }
+
+    private T[] createArray(T value, int arraySize) {
+        if (value == null && elements != null && elements.length > 0) {
+            value = elements[0];
+        }
+        return (T[]) Array.newInstance((value != null)
+                            ? value.getClass()
+                            : Object.class, arraySize);
+    }
```

Comment:
I can't get idea why simple things are such over-engeneered.
Why we create our list without inner array and wait for add() method to be called to make it.
Add constructor and make methods simplier:
```suggestion
    private void addCellForNewElement() {
        size++;
        if (size > elements.length) {
            T[] newArray = (T[]) Object[(int) (elements.length * GROW_FACTOR)];
            System.arraycopy(elements, 0, newArray, 0, elements.length);
            elements = newArray;
        }
    }
    
    //and add constructor to the top
    public ArrayList() {
        elements = (T[]) Object[INITIAL_SIZE];
    }
```

Reply:
Well, your mate mentioned I have to use T objects, not "Object" type.
So, what I did was just an approach to get an actual T type object array when it is possible.
In other words, with String objects, it created a real String[] array behind the scenes. Not an Object[] one.
Actually, my original version included the way you mentioned except for the type of elements array.
It was Object[], not T[]. Having it as a T[] initialized by Object[] doesn't look better than the Object[] elements array.
Not sure what was the idea to tell me to use T[] without initializing it with T[]...
https://github.com/mate-academy/jv-array-list/pull/1031#discussion_r1257294590

Any ideas?